source/extensions/quic_listeners/quiche/platform/quic_mem_slice_impl.cc

```diff
@@ -0,0 +1,51 @@
+// NOLINT(namespace-envoy)
+//
+// This file is part of the QUICHE platform implementation, and is not to be
+// consumed or referenced directly by other Envoy code. It serves purely as a
+// porting layer for QUICHE.
+
+#include "extensions/quic_listeners/quiche/platform/quic_mem_slice_impl.h"
+
+#include "common/common/assert.h"
+
+namespace quic {
+
+Envoy::Buffer::BufferFragmentImpl*
+QuicMemSliceImpl::allocateBufferAndFragment(QuicBufferAllocator* allocator, size_t length) {
+  auto bundle = reinterpret_cast<BufferFragmentBundle*>(
+      allocator->New(sizeof(BufferFragmentBundle) + length));
+  Envoy::Buffer::BufferFragmentImpl& fragment = bundle->fragment_with_padding_.fragment_;
+  return new (&fragment) Envoy::Buffer::BufferFragmentImpl(
```

Comment:
This looks very concerning. I think that Envoy::Buffer::BufferFragmentImpl should be more involved in how it gets allocated.

I see what you are trying to do -- reserve enough space for it and embed it inline in another data structure, but this looks very much to me like an abstraction boundary violation.

Note also there are two alternative implementations of Envoy Buffers. Does this work for both of them?

Reply:
Both SliceDeque and evbuffer implementations of Envoy::BufferImpl take a BufferFragmentImpl and release it via call back. How the fragment is allocated and deallocated is hidden from them. I couldn't see how it wouldn't support one of SliceDeque and evbuffer.

Reply:
it's worth cross-linking https://github.com/envoyproxy/envoy/pull/6732 which is also doing interesting things with buffers. I'm wondering if anything can be improved in the buffer API to make it easier to reason about call-sites.